Add gMSA documentation (#190)
rancher/windows: docs/active-directory/core-concepts/01-active-directory-details-and-definitions.md

# Active Directory Details and Definitions 

### Active Directory Domain Services

> Active Directory Domain Services (AD DS, or more simply just AD for Active Directory) is a storage and management

Reviewer: link to active directory docs

> Active Directory Domain Services (AD DS, or more simply just AD for Active Directory) is a storage and management
> mechanism for everything related to a collection of resources within a computer network. This includes users, files,
> network configurations, and devices like physical computers and printers. Administrators can modify the Active Directory
> and remotely manage all the components currently tracked at once. In its simplest terms, **AD is used to logically
> represent an organizations resources and users and enforce policies and behaviors across them.**

Reviewer: ```suggestion
Active Directory Domain Services (often informally referred to as AD DS or AD) is a storage and management mechanism for everything related to a collection of resources within a computer network, including users, files, network configurations, and devices like physical computers and printers. 

In its simplest terms, Active Directory is used to **create a logical representation of an organization's resources and users**, which allows them to **remotely manage these resources** and **enforce policies and behaviors across them**.
```

Reviewer: in general, there seem to be manually added newlines! Instead, I think a better approach is to create line breaks between sentences (instead of defining paragraphs) and let sentences stay on the same line.

> AD Includes the following features

Reviewer: ```suggestion
AD Includes the following features:
```

Reviewer: In general, I think it's easier to read the docs if we just replace AD -> Active Directory.

- Data schemas which enforce particular requirements onto AD data
- A global catalog of all data
- Query and Indexing mechanisms for finding and publishing specific data

> - Database replication to various Domain Controllers (DC)

Reviewer: This is a little dense, maybe this should be converted into a paragraph?

> ### AD Storage Structure

Reviewer: It's assumed these docs are all on AD, so no need to mention AD in the heading.

[From MS Docs](https://learn.microsoft.com/en-us/previous-versions/windows/it-pro/windows-server-2003/cc759186(v=ws.10))

“*Administrators use **Active Directory** to store and organize objects on a network (such as users, computers,
devices, and so on) into a secure hierarchical containment structure that is known as the **logical structure**.
Although the **logical structure** of **Active Directory** is a hierarchical organization of all users, computers,
and other physical resources, the **forest** and **domain** form the basis of the logical structure. **Forests**,
which are the **security boundaries of the logical structure**, can be structured to provide data and service autonomy
and isolation in an organization in ways that can both reflect site and group identities and remove dependencies on the

> physical topology.”*

Reviewer: Use a block quote, i.e.

> This is a block quote

Reviewer: Also, if they wanted to read this, couldn't someone just click on the MS docs linked? In this section, I think it's better to link to the docs and instead split this up into two subsections:
1. A terminology section that just covers each term and a brief description
2. A "dumbed down" version of the docs, leveraging an analogy that explains the relationship between all of the above terms

i.e.

---

Since Microsoft assumes that an organizational's entire view of all of the resources owned by it is tracked within the Active Directory, it supports the ability to form logical groupings of those resources (i.e. **organizational units**) within the Active Directory that represent discrete segments of your organization. These discrete segments can be managed independently by administrators, which can be used to apply security and organizational policies to groups of resources as a whole.

As an analogy, if you think of Active Directory as a supermarket, **forests** represent different divisions of the supermarket (i.e. Home, Technology, Grocery, etc.) whereas **domains** are like aisles within a specific division (i.e. you may have a aisle within Grocery dedicated to breakfast items).

Active Directory represents these entities and their relationships using a flat database file that encodes a tree of relationships. For example, in our supermarket analogy:

```
Grocery/
  Breakfast Foods/
  Baking Goods/
  etc.
Home/
  etc.
Technology/
  etc.
```

---

Once you have such an analogy as above to give a strong image of what we're talking about, the more complicated description becomes more accessible to read!

Concepts like the fact that multiple forests exist in a single ActiveDirectory also more naturally flow from that analogy.

- Active Directory Forest
    - The highest level in the hierarchy. A forest represents a collection of Directories. It is a security boundary, 
    which gives administrators complete control and access to all the information inside both the Forest and all
    Domains within the Forest. Multiple Forests may exist at a time.
- Domain
    - Domains partition the information that is stored inside the directory into smaller portions so that the
  information can be more easily stored on various domain controllers and so that administrators have a greater
  degree of control over replication. Data that is stored in the Active Directory is replicated throughout the forest from
  one domain controller to another. Some data that is relevant to the entire forest is replicated to all domain
  controllers. Other data that is relevant only to a specific domain is replicated only to domain controllers in
  that particular domain. A good domain design makes it possible to implement an efficient replication topology.
  This is important because it enables administrators to manage the flow of data across the network, that is, to

>   control how much data is replicated and where that replication traffic takes place.

Reviewer: This entire concept may be deserving of its own section; I would honestly say we should move the deeper explanation of what domains are to a completely different doc.

- Organizational Unit (OU)
    - OU’s provide a way to group resources so that they can be managed as a single unit. This allows admins to apply
group policies across many computers or control access of resources for many users. OUs can also be delegated to 

> specific admins, so that they only have control over their subset of resources.

Reviewer: I'm not quite sure from this description what OUs are. Are they a concept that exists on a forest / domain level or can you group together any set of resources in an AD across forests or domains?

Would be great if this can be part of the analogy recommended above.

Other core components of an Active Directory are

- DNS Support
    - In Active Directory, DNS is the means by which directory clients locate, or discover, domain controllers. 
      The primary components of the architecture for DNS support of Active Directory include the domain controller
      Locator, Active Directory domain names in DNS, and Active Directory DNS objects.
    - Every Active Directory domain has a DNS domain name (for example, [cohovineyard.com](http://cohovineyard.com/)),
      and every domain joined computer has a DNS name (for example, 
      [server1.cohovineyard.com](http://server1.cohovineyard.com/)). Architecturally, domains and computers are

>       represented both as objects in Active Directory and as nodes in DNS.

Reviewer: Link to the wikipedia page about DNS. Also, I think this doesn't necessarily need a large explanation.

The fact that every domain has its own DNS domain name should be moved to docs specifically built around domains.

Don't think you need to get into the specifics that they are represented as nodes.

- Active Directory Schema
    - The schema contains definitions of objects which can store information in the Active Directory. It provides the 
      attributes associated with particular resources.
    - Different Windows versions will have different schema versions, when different versions of windows are used to 

>       comprise the Active Directory server they will agree upon a shared Schema.

Reviewer: What format is this schema stored in? XAML? or something else?

Also think this should be discussed in another document. The fact that AD supports data validation for user inputs seems like a feature, not a core concept.

Author: The active directory schema is effectively the schema of the database, I haven't found documentation which point to a particular file format or a way to put the schema into a human readable text format. Users shouldn't be changing the schema anyway. Agree that it can be moved to section focusing more specifically on domains

- The Data store
    - A data store comprises three components. The first being a collection of interfaces used to communicate with the
      store, the second being the underlying services and logic which perform CRUD on the store, and the third is the
      literal data. Each data store is represented as a single file replicated across all domain controllers.
    - Proper logical structuring of an Active Directory will ensure that Domain Controllers will only have access to the
      data required for their domain, keeping replication quick and simple. Improper structuring can saturate the

>       network with replication data. Properly structuring an on premise Active Directory is an art.

Reviewer: I think this should be in a section about data replication / disaster recovery. Also does not feel like a core concept of AD, seems like a specific feature for admins.

### The Components and Interfaces of Active Directory Storage

There are several storage components that come together to form the data handling of Active Directory

> The core interfaces used to communicate with the storage layer are

Reviewer: ```suggestion
The core interfaces used to communicate with the storage layer are:
```

- LDAP, Lightweight Directory Access Protocol
- Replication within the domain controller management interface
- A Messaging API
- A Security Accounts Manager (SAM)

The core components of the storage layer are

- Directory System Agent (DSA)
- The Database, typically represented as a single file replicated across all Domain Controllers
- Extensible Storage Engine (ESE)

> ![Data Store Architecture](../media/Untitled%201.png)

Reviewer: Again, I think this needs an analogy or some more specific explanation to make this more accessible.
